docs/extension.rst

```diff
@@ -0,0 +1,417 @@
+Extensions
+==========
+
+A brief overview of extensions can be found in :doc:`dependencies`. This 
+guide will go in depth on how to write an extension manifest and explain 
+the various properties used.
+
+The first step on writing or loading an extension is to define an extension
+point or use an existing one. The decision depends on where the extension
+should be made available.
+
+If for example it is a GTK or QT theme, it's best to define or use an
+extension point in the runtime, so that it can be made available for 
+every application utilising said runtime.
+
+If it is a specific tool eg. MPV plugins, which are only used by MPV, 
+it's best to define or use an extension point in the application.
+
+To see what are the available extension points in runtime or 
+applications, please see the corresponding flatpak or build
+manifest.
+
+Extension point
+---------------
+
+Now to define an extension point the ``add-extensions`` property can be 
+used in the manifest.
+
+This is a typical example of an application extension point.
+
+.. code-block:: yaml
+
+  id: org.flatpak.app
+  runtime: org.gnome.Platform
+  runtime-version: '45'
+  sdk: org.gnome.Sdk
+  command: foo
+  add-extensions:
+    org.flatpak.app.plugin:
+      version: '1.0'
+      directory: extension_directory
+      add-ld-path: lib
+      merge-dirs: plug-ins;help
+      subdirectories: true
+      no-autodownload: true
+      autodelete: true
+  cleanup-commands:
+    - mkdir -p ${FLATPAK_DEST}/extension_directory
```

Comment:
cleanup-commands can be anything that runs during the cleanup phase. Runtimes run various integration commands here.

This is the only way afaik to ensure that something runs after the entirety of the build step in the manifest is done.

Reply:
`cleanup-commands`. It's illogical.

They go in a `post-install`.

Reply:
And just to clarify this can be done a *anytime*. It just need to exist in the final flatpak.

Reply:
post-install means some module/command can rm the empty directory from the flatpak if it is not the last module, cleanup-command prevents that

Reply:
> post-install means some module/command can rm the empty directory from the flatpak if it is not the last module/command

No. Only the top-level cleanup or the module that created it can do that.
And like anything else. They can cleanup the binary too.

Reply:
> No. Only the top-level cleanup or the module that created it can do that.

I don't think that's true? Inside build-commands, the modules have access to sandbox filesystem and anything previously built/installed.

<details>
<summary> example </summary>

```yaml
id: org.flatpak.Hello
runtime: org.freedesktop.Platform
runtime-version: '23.08'
sdk: org.freedesktop.Sdk
command: hello
modules:

  - name: hello
    buildsystem: simple
    build-commands:
      - install -Dm755 hello.sh /app/bin/hello
    sources:
      - type: script
        dest-filename: hello.sh
        commands:
          - echo "Hello world"
    post-install:
      - mkdir -p /app/plugins

  - name: second
    buildsystem: simple
    build-commands:
      - rmdir -v /app/plugins
```

Reply:
Point is it doesn't belong to `cleanup-commands`. It's not a cleanup. It's a setup. We are writing a doc and the doc should be correct. I'm pretty sure if you the person that put it in there did it because it was done elsewhere.

Reply:
ideally the declaration of the extension point should be enough to flatpak-builder to create it, but it doesn't.

Reply:
> I'm pretty sure if you the person that put it in there did it because it was done elsewhere.

You ignored multiple reasons I provided above about why I prefer it to `post-install`...

Reply:
I didn't see any valid one.

Your hypothetical manifest could just as well remove the binaries. or a depenency. It's irrelevant to the logic that something to "setup" doesn't belong in a cleanup.

Reply:
I also see no valid reason to change it. I added that it can be done in post-install, so I cosidered it done.

Reply:
it about at making the documentation correct. that's the whole reason. Hence changing it is needed.

Reply:
It's already correct and works. Only you think that cleanup-commands is meant only to run rm. This is mentioned nowhere and the usage here consistent with what it can do and previous docs https://docs.flatpak.org/en/latest/manifests.html#cleanup